src/Compilers/CSharp/Portable/Lowering/Instrumentation/Instrumenter.cs

```diff
@@ -256,7 +256,7 @@ public virtual BoundStatement InstrumentPatternSwitchStatement(BoundPatternSwitc
         public virtual BoundStatement InstrumentPatternSwitchWhenClauseConditionalGotoBody(BoundExpression original, BoundStatement ifConditionGotoBody)
         {
             Debug.Assert(!original.WasCompilerGenerated);
-            Debug.Assert(original.Syntax.Parent.Kind() == SyntaxKind.WhenClause);
+            Debug.Assert(original.Syntax.FirstAncestorOrSelf<WhenClauseSyntax>() != null);
```

Comment:
>Debug.Assert(original.Syntax.FirstAncestorOrSelf<WhenClauseSyntax>() != null); [](start = 12, length = 78)

I think the assert can be stronger, i.e. it should skip only enclosing parens. Otherwise, it will not catch an incomplete expression.